p2p/src/peers/mod.rs

```diff
@@ -1,85 +1,282 @@
+    }
+
+    /// Contains for new bucket
+    pub fn new_bucket_contains(&self, index: u16) -> bool {
+        self.new_bucket.contains_key(&index)
+    }
+
+    /// Contains for tried bucket
+    pub fn tried_bucket_contains(&self, index: u16) -> bool {
+        self.tried_bucket.contains_key(&index)
+    }
+
+    /// Returns the timestamp of a specific slot in the new addresses bucket
+    pub fn new_bucket_get_timestamp(&self, index: u16) -> Option<i64> {
+        self.new_bucket.get(&index).map(|p| p.timestamp)
+    }
+
+    /// Returns the timestamp of a specific slot in the tried addresses bucket
+    pub fn tried_bucket_get_timestamp(&self, index: u16) -> Option<i64> {
+        self.tried_bucket.get(&index).map(|p| p.timestamp)
+    }
+
+    /// Returns the timestamp of a specific slot in the new addresses bucket
+    pub fn new_bucket_get_address(&self, index: u16) -> Option<SocketAddr> {
+        self.new_bucket.get(&index).map(|p| p.address)
+    }
+
+    /// Returns the timestamp of a specific slot in the tried addresses bucket
+    pub fn tried_bucket_get_address(&self, index: u16) -> Option<SocketAddr> {
+        self.tried_bucket.get(&index).map(|p| p.address)
+    }
+
+    /// Add multiple peer addresses and save timestamp in the new addresses bucket
     /// If an address did already exist, it gets overwritten
     /// Returns all the overwritten addresses
-    pub fn add(&mut self, addrs: Vec<SocketAddr>) -> Result<Vec<SocketAddr>, failure::Error> {
+    pub fn add_to_new(
+        &mut self,
+        addrs: Vec<SocketAddr>,
+        src_address: SocketAddr,
+    ) -> Result<Vec<SocketAddr>, failure::Error> {
         // Insert address
         // Note: if the peer address exists, the peer info will be overwritten
-        Ok(addrs
+        let result = addrs
             .into_iter()
             // Filter out unspecified addresses (aka 0.0.0.0)
             .filter(|address| !address.ip().is_unspecified())
             .filter_map(|address| {
-                self.peers
+                let index = self.new_bucket_index(&address, &src_address);
+
+                self.new_bucket
                     .insert(
-                        address,
+                        index,
                         PeerInfo {
                             address,
-                            _timestamp: get_timestamp(), //msg.timestamp,
+                            timestamp: get_timestamp(), //msg.timestamp,
                         },
                     )
                     .map(|v| v.address)
             })
-            .collect())
+            .collect();
+
+        Ok(result)
     }
 
-    /// Remove a peer given an address
+    /// Add multiple peer addresses and save timestamp in the tried addresses bucket
+    /// If an address did already exist, it will be checked and gets overwritten
+    /// in case of it would not be valid
+    /// Returns all the overwritten or rejected addresses
```

Comment:
Shouldn't peers be checked before getting overwritten? This implementation will always overwrite the peers in tried, is that the expected behavior or will it change in the future?